File: backend/src/controllers/admin.controller.ts
import type { Request, Response } from "express";
import { adminService } from "../services/admin.service.js";
import { ApiResponse } from "../utils/api-response.js";

const param = (request: Request, name: string): string => request.params[name] as string;

export class AdminController {
  async dashboard(request: Request, response: Response) {
    const stats = await adminService.getDashboardStats();
    response.json(new ApiResponse(200, stats, "Dashboard stats retrieved"));
  }

  async listUsers(request: Request, response: Response) {
    const result = await adminService.listUsers(request.query as Record<string, string>);
    response.json(new ApiResponse(200, result, "Users retrieved"));
  }

  async getUser(request: Request, response: Response) {
    const user = await adminService.getUser(param(request, "userId"));
    response.json(new ApiResponse(200, user, "User retrieved"));
  }

  async banUser(request: Request, response: Response) {
    const user = await adminService.banUser(request.user!.id, param(request, "userId"));
    response.json(new ApiResponse(200, user, "User banned"));
  }

  async unbanUser(request: Request, response: Response) {
    const user = await adminService.unbanUser(request.user!.id, param(request, "userId"));
    response.json(new ApiResponse(200, user, "User unbanned"));
  }

  async activateUser(request: Request, response: Response) {
    const user = await adminService.activateUser(request.user!.id, param(request, "userId"));
    response.json(new ApiResponse(200, user, "User activated"));
  }

  async suspendUser(request: Request, response: Response) {
    const user = await adminService.suspendUser(request.user!.id, param(request, "userId"));
    response.json(new ApiResponse(200, user, "User suspended"));
  }

  async deleteUser(request: Request, response: Response) {
    await adminService.deleteUser(request.user!.id, param(request, "userId"));
    response.json(new ApiResponse(200, null, "User deleted"));
  }

  async listCommunities(request: Request, response: Response) {
    const result = await adminService.listCommunities(request.query as Record<string, string>);
    response.json(new ApiResponse(200, result, "Communities retrieved"));
  }

  async deleteCommunity(request: Request, response: Response) {
    await adminService.deleteCommunity(request.user!.id, param(request, "communityId"));
    response.json(new ApiResponse(200, null, "Community deleted"));
  }

  async listResources(request: Request, response: Response) {
    const result = await adminService.listResources(request.query as Record<string, string>);
    response.json(new ApiResponse(200, result, "Resources retrieved"));
  }

  async deleteResource(request: Request, response: Response) {
    await adminService.deleteResource(request.user!.id, param(request, "resourceId"));
    response.json(new ApiResponse(200, null, "Resource deleted"));
  }

  async listReports(request: Request, response: Response) {
    const result = await adminService.listReports(request.query as Record<string, string>);
    response.json(new ApiResponse(200, result, "Reports retrieved"));
  }

  async deleteMessage(request: Request, response: Response) {
    await adminService.deleteMessage(request.user!.id, param(request, "messageId"));
    response.json(new ApiResponse(200, null, "Message deleted"));
  }

  async deleteDirectMessage(request: Request, response: Response) {
    await adminService.deleteDirectMessage(request.user!.id, param(request, "messageId"));
    response.json(new ApiResponse(200, null, "Direct message deleted"));
  }

  async reviewReport(request: Request, response: Response) {
    const report = await adminService.reviewReport(request.user!.id, param(request, "reportId"), request.body);
    response.json(new ApiResponse(200, report, "Report reviewed"));
  }
}

export const adminController = new AdminController();

File: backend-java/src/main/java/com/studyconnect/backend/controller/AdminController.java
package com.studyconnect.backend.controller;

import com.studyconnect.backend.dto.ApiResponse;
import com.studyconnect.backend.dto.admin.ReviewReportRequest;
import com.studyconnect.backend.service.AdminService;
import com.studyconnect.backend.util.SecurityUtil;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.bind.annotation.RequestBody;

@RestController
@Validated
@PreAuthorize("hasRole('ADMIN')")
@RequestMapping("/api/admin")
public class AdminController {

    private final AdminService adminService;

    public AdminController(AdminService adminService) {
        this.adminService = adminService;
    }

    @GetMapping("/dashboard")
    public ApiResponse<?> dashboard() {
        return ApiResponse.success(adminService.getDashboardStats(), "Dashboard stats retrieved");
    }

    @GetMapping("/users")
    public ApiResponse<?> listUsers(
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String role,
            @RequestParam(required = false) String status,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit) {
        return ApiResponse.success(adminService.listUsers(search, role, status, page, limit), "Users retrieved");
    }

    @GetMapping("/users/{userId}")
    public ApiResponse<?> getUser(@PathVariable String userId) {
        return ApiResponse.success(adminService.getUser(userId), "User retrieved");
    }

    @PatchMapping("/users/{userId}/ban")
    public ApiResponse<?> banUser(@PathVariable String userId, Authentication authentication) {
        return ApiResponse.success(adminService.banUser(SecurityUtil.currentUserId(authentication), userId), "User banned");
    }

    @PatchMapping("/users/{userId}/unban")
    public ApiResponse<?> unbanUser(@PathVariable String userId, Authentication authentication) {
        return ApiResponse.success(adminService.unbanUser(SecurityUtil.currentUserId(authentication), userId), "User unbanned");
    }

    @PatchMapping("/users/{userId}/activate")
    public ApiResponse<?> activateUser(@PathVariable String userId, Authentication authentication) {
        return ApiResponse.success(adminService.activateUser(SecurityUtil.currentUserId(authentication), userId), "User activated");
    }

    @PatchMapping("/users/{userId}/suspend")
    public ApiResponse<?> suspendUser(@PathVariable String userId, Authentication authentication) {
        return ApiResponse.success(adminService.suspendUser(SecurityUtil.currentUserId(authentication), userId), "User suspended");
    }

    @DeleteMapping("/users/{userId}")
    public ApiResponse<?> deleteUser(@PathVariable String userId, Authentication authentication) {
        adminService.deleteUser(SecurityUtil.currentUserId(authentication), userId);
        return ApiResponse.success(null, "User deleted");
    }

    @GetMapping("/communities")
    public ApiResponse<?> listCommunities(
            @RequestParam(required = false) String search,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit) {
        return ApiResponse.success(adminService.listCommunities(search, page, limit), "Communities retrieved");
    }

    @DeleteMapping("/communities/{communityId}")
    public ApiResponse<?> deleteCommunity(@PathVariable String communityId, Authentication authentication) {
        adminService.deleteCommunity(SecurityUtil.currentUserId(authentication), communityId);
        return ApiResponse.success(null, "Community deleted");
    }

    @GetMapping("/resources")
    public ApiResponse<?> listResources(
            @RequestParam(required = false) String search,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit) {
        return ApiResponse.success(adminService.listResources(search, page, limit), "Resources retrieved");
    }

    @DeleteMapping("/resources/{resourceId}")
    public ApiResponse<?> deleteResource(@PathVariable String resourceId, Authentication authentication) {
        adminService.deleteResource(SecurityUtil.currentUserId(authentication), resourceId);
        return ApiResponse.success(null, "Resource deleted");
    }

    @GetMapping("/reports")
    public ApiResponse<?> listReports(
            @RequestParam(required = false) String status,
            @RequestParam(required = false) String targetType,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit) {
        return ApiResponse.success(adminService.listReports(status, targetType, page, limit), "Reports retrieved");
    }

    @PatchMapping("/reports/{reportId}")
    public ApiResponse<?> reviewReport(
            @PathVariable String reportId,
            @Valid @RequestBody ReviewReportRequest request,
            Authentication authentication) {
        return ApiResponse.success(
                adminService.reviewReport(SecurityUtil.currentUserId(authentication), reportId, request),
                "Report reviewed"
        );
    }

    @DeleteMapping("/messages/{messageId}")
    public ApiResponse<?> deleteMessage(@PathVariable String messageId, Authentication authentication) {
        adminService.deleteMessage(SecurityUtil.currentUserId(authentication), messageId);
        return ApiResponse.success(null, "Message deleted");
    }

    @DeleteMapping("/direct-messages/{messageId}")
    public ApiResponse<?> deleteDirectMessage(@PathVariable String messageId, Authentication authentication) {
        adminService.deleteDirectMessage(SecurityUtil.currentUserId(authentication), messageId);
        return ApiResponse.success(null, "Direct message deleted");
    }
}
